components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -19,58 +19,83 @@ use exonum_crypto::Hash;
 
 use super::{super::BinaryValue, key::ProofListKey, HashTag};
 
-/// Encapsulates a proof of absence for `ProofListIndex`.
+/// An enum that represents a proof of existence/absence for a proof list elements.
 ///
 /// Proof of absence for an element with the specified index consists of
 /// `merkle_root` of `ProofListIndex` and `length` of the list.
 ///
-/// Element with `index` is absent in the list with provided `length`
-/// and `merkle_root` when two conditions are met:
+/// Element with `index` is absent in the list with provided
+/// `merkle_root` when two conditions are met:
 /// ```text
 /// 1. list_hash == sha256( HashTag::List || length || merkle_root )
 /// 2. index > length
 /// ```
 ///
 /// In case of a range proof this rule applies to the whole range.
-#[derive(Debug, PartialEq, Clone, Eq)]
-pub struct ProofOfAbsence {
+#[derive(Debug, Clone, PartialEq, Eq)]
+//TODO: serialize/deserialize with serde derive [ECR-3222]
+pub enum ProofVariant<V> {
+    /// A branch of proof in which both children contain requested elements.
+    Full(Box<ProofVariant<V>>, Box<ProofVariant<V>>),
+    /// A branch of proof in which only the left child contains requested elements.
+    Left(Box<ProofVariant<V>>, Option<Hash>),
+    /// A branch of proof in which only the right child contains requested elements.
+    Right(Hash, Box<ProofVariant<V>>),
+    /// A leaf of the proof with the requested element.
+    Leaf(V),
+    /// Proof of absence of an element with the specified index.
+    Absent(Hash),
```

Comment:
Am I correct that `Absent` variant can occur only as the direct value of `ListProof.proof`? In this case, the variant doc comment seems misleading: the variant does not proof an absence of a specific element, but rather a proof of absence of a range. (Arguably, even this would be incorrect; see the comment about partially valid ranges above.)